src/pages/VeSDL/index.tsx

```diff
@@ -11,179 +11,173 @@ import {
-      (MULTIPLIER * leftTime) / MAXTIME,
-    ).toFixed()
+  const currentTimestamp = getUnixTime(new Date())
+  const unlockDateOrLockEnd = unlockDate || lockEnd
+  const expireTimestamp =
+    unlockDateOrLockEnd && !isNaN(unlockDateOrLockEnd.valueOf())
+      ? getUnixTime(unlockDateOrLockEnd)
+      : undefined
+  const totalAmount =
+    !sdlTokenValue.isZero() || unlockDate
+      ? sdlTokenValue.add(lockedSDLVal)
+      : Zero
 
-    const penaltyAmount = lockedSDLVal
-      .mul(BigNumber.from(penaltyRatio))
-      .div(BigNumber.from(MULTIPLIER))
-    return formatUnits(penaltyAmount)
-  }
+  // Calculate estimated veSdl amount
+  const estLockAmt = expireTimestamp
+    ? totalAmount
+        .mul(
+          BigNumber.from(expireTimestamp).sub(BigNumber.from(currentTimestamp)),
+        )
+        .div(BigNumber.from(MAXTIME))
+    : Zero
+
+  // Calculate penalty Ratio and penalty amount
+  const leftTimeForUnlock = lockEnd && getUnixTime(lockEnd) - currentTimestamp
+  const penaltyAmount = leftTimeForUnlock
+    ? minBigNumber(
+        lockedSDLVal.mul(BigNumber.from(3)).div(BigNumber.from(4)),
+        lockedSDLVal
+          .mul(BigNumber.from(leftTimeForUnlock))
+          .div(BigNumber.from(MAXTIME)),
+      )
+    : Zero
 
   const handleLock = async () => {
+    const unlockTimeStamp = unlockDate ? getUnixTime(unlockDate) : null
+    const hasLockedSDL = !lockedSDLVal.isZero()
+    const shouldCreateLock =
+      !hasLockedSDL && sdlTokenValue.gt(Zero) && unlockTimeStamp
+    const shouldIncreaseAmoutAndLockEnd =
+      hasLockedSDL && sdlTokenValue.gt(Zero) && unlockTimeStamp
+    const shouldIncreaseAmount =
+      hasLockedSDL && sdlTokenValue.gt(Zero) && !unlockTimeStamp
+    const shouldIncreaseLockEnd =
+      hasLockedSDL && sdlTokenValue.isZero() && unlockTimeStamp
+
     if (!account || !chainId || !votingEscrowContract?.address) return
     try {
-      const unlockTimeStamp = unlockDate ? getUnixTime(unlockDate) : null
-      if (sdlTokenValue.gt(Zero))
-        await sdlContract?.approve(votingEscrowContract?.address, sdlTokenValue)
+      if (sdlTokenValue.gt(Zero)) {
+        await checkAndApproveTokenForTrade(
+          sdlContract,
+          votingEscrowContract.address,
+          account,
+          sdlTokenValue,
+          infiniteApproval,
+        )
+      }
 
-      if (sdlTokenValue.gt(Zero) && unlockTimeStamp) {
-        if (lockedSDLVal.isZero()) {
-          const txn = await votingEscrowContract?.create_lock(
-            sdlTokenValue,
-            unlockTimeStamp,
-          )
-          await txn.wait()
-        } else {
-          const txnIncreaseAmount = await votingEscrowContract.increase_amount(
-            sdlTokenValue,
+      if (shouldCreateLock && unlockTimeStamp) {
+        const txn = await votingEscrowContract?.create_lock(
+          sdlTokenValue,
+          unlockTimeStamp,
+        )
+        await txn.wait()
+        enqueueToast("success", "Locked SDL")
```

Comment:
can you use `enqueuePromiseToast` here? 

```
await enqueuePromiseToast(chainId, transaction.wait(), "createLock")
```

you will probably have to create a case for "createLock" to get the message correct to the user in the toast.